### ISSUE TITLE
Models storage and distribution 

### ISSUE THREAD

**Issue by VBPXKSMI** [Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab#159]
Edited everything because Github is uttershit.

**Comment by VBPXKSMI** [comment]
So, we start to have quite some models now, but we seriously lack of a decent way to store/distribute them. Torrents aren't viable long term (as seen with pytorch) unless there's always one occurrency on a seedbox server (always available, big seed speed, but paid). Mega has a size limit of 5Gb so any larger model won't be able to download. Google drive could be a solution, but I know they tends to block the files that gets too much downloaded too quickly.

If we want a colab that works correctly, we need to find a storage that

•	doesn't rely on the people who already have the file
•	have a decent number of download of one file (unlimited would be ideal)
•	doesn't limit the download speed, or at a pretty decent download speed (no 100Kb shit)

I was considering dropbox, but I don't know its limitations. an other solution i came with was store them directly in our github (since the distribution problem is mainly a colab issue), but I don't know if it would be possible.

**Comment by VBPXKSMI** [comment]
> an other solution i came with was store them directly in our github, but I don't know if it would be possible.

Don't know about how feasible that is, specially when we get more models.

One thing I found when looking at the mormon GitHub page was this issue:
> github (dot)com (slash) (AIDungeon) (slash) (AIDungeon) (slash) (issues) (slash) (87)

An then this pull request addresing it:
>  github (dot) com (slash) (AIDungeon) (slash) (AIDungeon) (slash) (pull) (slash) (107)

Perhaps, maybe we can do something similar and force everyone that uses the Colab to seed the torrent? For a fixed (or not) amount of time?
Honestly, I think that would be the best solution, torrent is just too convenient to use and it would prevent the files from getting lost forever.

**Comment by VBPXKSMI** [comment]
it's possible by modifying aria2 parameter in the actual code, but it prevents the cell to go any further until aria said "yep, your seed is done!"

edit : after looking the script, that's basically what the scripts in thadunge and clover's repo do, the scripts I bypassed in the new code. the thing is, that way, it seeds only while the cell is running. and as said, it'll prevent it to go any further until ratio or seed time is not fulfilled, with my code. the concept of the script is to hook aria's output to /dev/null to make it disappear, which allows to run the game... I know parameters can be passed to hook aria2 when the download is completed, maybe I can mess with that to have it run the game while it stays in background to seed. gonna investigate that further, but still, as I mentionned, torrents, even if they are convenient, are not viable long term. they'll die if no one seed themunless someone put an occurence on a seedbox server. for now, it's good, but after a while, it'll be problematic

**Comment by VBPXKSMI** [comment]
Well, it was an idea. Also torrent should be fine for now but I don't know if other host sites are that viable but yes, we need to investigate further.

**Comment by AtaIanta** [comment]
If somehow no one would seed pytorch models, can't you guys use the convert script in clover edition to convert tensorflow model?

**Comment by machiavel23** [comment]
well, I actually realized that the models in the mega are quite small (under 3Gb) and I should be able to make the cell download them. The script of clover can also be used by the cell to convert them, meaning we only need the gpt2 models putted in that mega. currently rewriting the cell for that. Only thing I'd like is the vanilla model_v5 to be uploaded on the mega, to ave everything at the same place

**Comment by VBPXKSMI** [comment]
Vanilla models (as in the untouched GPT-2 ones) + AI Dungeon base models are here:
>EDIT: link removed
Meganon will have to upload them himself, alongside the texts I provided earlier.

Other things that still need to be put there in the mega folder may be found here:
>EDIT: link removed

**Comment by machiavel23** [comment]
thanks. even without it being in the same mega for now will help me do what i want

**Comment by lordpidey** [comment]
An option, if we have someone willing to foot some of the bill, and can get the model under 5GB, is to use AWS S3 to be a permanent seedbox.   It's little known, but AWS S3 does support torrent based distribution for files under 5GB.

**Comment by machiavel23** [comment]
isn't that what (((HE))) did? besides, I doubt anyone in the thread would be willing to pay for that. as long as the models can be kept under 5Gb, we can perfectly use mega

**Comment by lordpidey** [comment]
No, that's not quite what was done.

From my understanding, he was using S3 (or something similar) to directly host for an initial time period, then switched to a pure torrent model, utilizing a standard seedbox.

I'm proposing using S3 as a seedbox.

**Comment by machiavel23** [comment]
I've checked s3... it's a paid solution. Nobody would accept to pay that, even if it's a small fee :/

**Comment by Rioghasarig** [comment]
I think we can use Google Storage.  

The original ran into problems because he thousands people tried to download the data all of sudden, so it was very costly. But you can set things up in such away that the "requester pays". This will protect the host form paying an enormous amount if thousands of people suddenly downloaded it. And the price for each person to download it will be literal pennies. And since google literally gives hundreds of dollars worth of GCP credits to new users it actually won't cost them anything. 

So, yeah, I think I'll do this part myself.

**Comment by Rioghasarig** [comment]
Okay, so currently I've got a bucket containing the pytorch clover edition model here: 
https://storage.googleapis.com/clover-edition/model_v5_pytorch.zip (EDIT: Just remembered this version of the game uses the tensorflow model. Will be adding that soon)

Right now it's free and easy for anyone to access. If the costs start getting to much though it'll be taken down and I'll fall back to the "requester pays" solution described above. I don't think it'll be a problem but  if it is I'll also post instructions on how to access the bucket then.

**Comment by machiavel23** [comment]
I'll do a test cell soon using your link to retrieve the model file (will be mostly a clone of the current cell with the model download part adapted) and drop the link of it in the thread. we'll see how people react to it

**Comment by Rioghasarig** [comment]
So I've uploaded the tensorflow model. It's available at

https://storage.googleapis.com/clover-edition/model_v5.zip 

Europeans should use  https://storage.googleapis.com/clover-europe/model_v5.zip for faster download speed.

**Comment by machiavel23** [comment]
so I did a few tests on colab:

pytorch
retrieval takes around 2 minutes, unzipping takes around 4 minutes.

model_v5
retrieval takes around 2 minutes, unzipping send error message:
```
warning [model_v5.zip]:  1484409052 extra bytes at beginning or within zipfile
  (attempting to process anyway)
error [model_v5.zip]:  start of central directory not found;
  zipfile corrupt.
  (please check that you have transferred or created the zipfile in the
  appropriate BINARY mode and that you have compiled UnZip properly)
```

model_v5 europe
retrieval failed because of access forbidden:
```
Resolving storage.googleapis.com (storage.googleapis.com)... 108.177.111.128, 2607:f8b0:4001:c12::80
Connecting to storage.googleapis.com (storage.googleapis.com)|108.177.111.128|:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
2020-01-19 03:22:24 ERROR 403: Forbidden.
```
conclusion

The europe version is mostly useful for the offline versions honestly. On colab, it's a google to google transfer so it's not useful, and I guess it's also why it throws a 403.

I'm more concerned about the "corrupted" zip of tensorflow model, I think it might be to big to be correctly zipped or downloaded with wget (downloader used on colab)

also, I'd like to ask @Rioghasarig if it was possible to have them directly in folder instead of zip? not mandatory, it's just it would be more practical for colab, but I can easily deal with zips if there's no other way.

all in all, it seems to be a good solution (fast at least), and if we use it only for the model retrieval on the setup phase, it shouldn't cost too high. The particularly interesting thing is if you could do that with all currently available models from the mega folder (not requesting it obviously, but if you can take a look, and evaluate if it could be done, please let me know).

**Comment by VBPXKSMI** [comment]
>The particularly interesting thing is if you could do that with all currently available models from the mega folder

Adding to this, I will try to get in contact with MegAnon tomorrow so that all models are in his Mega, currently they are all around the place for my tastes. It shouldn't get past the 50GB free limit so it should be good, bad thing about it is that all models will be (well they are already) only available on Mega with its fucking stupid 5GB download limit.

**Comment by VBPXKSMI** [comment]
@Rioghasarig I want to add something, after seeing the issue you made in CloverAnon's repo.
I am thankful for what you are doing, but I also think we shouldn't exclusively rely on paid options. Too few of us would probably invest on this, otherwise we would be paying the mormon already.
Not saying this isn't a good option, just that long term a torrent (or another free host site) based method should be in place too, in case the paid method becomes unreliable.

Main problem I see with torrent is that new models added may be a pain in the ass to add to the already existing torrents (I think there is a certain way on adding new things to an already existing torrent, but I may be wrong since I didn't investigate on this too much before). Also another problem is that as time goes on, the amount of GB of models may get too big and probably noone will have them all on their PCs so seeding may become a bit of an issue (maybe).

**Comment by Rioghasarig** [comment]
@machiavel23  

Yeah I think I messed up a few things. Like I forgot to make the European bucket publicly accessible.  That should be fixed now. And I noticed there was an error when I uploaded the zip file. It restarted so I though there was no problem but maybe it got corrupted.  I'll reupload it when I get the chance.

I chose to use a zip file because I thought that would be more convenient for most people. You know, download one file instead of several.  I think this is the best way to store things for most people to access them. 

Also I don't think the colab should be set up to redownload the model every time. That'd be inefficient. There should be some permanent storage option or maybe an option to upload it from your PC if you've already downloaded it. 

And yeah I'm happy to host whatever models are out there. But before that I'm looking for a way to make browsing the models easy. Right now I only have a few models up but if there's a lot it could get confusing. I think I'll set up a webpage on github for that. 

@VBPXKSMI 
I really don't think we'll find a solution better than Google cloud storage and definitely not for free. First off, the cost of hosting the data is very small. It's like 2 cents per gb per month. So I can easily host a few hundred GB no problem. 

The real problem is network transfer costs. On google cloud storage transfer costs are 1 c per GB (if you download from the bucket in the right continent). So I can handle the model being downloaded a few thousand times per month.  But I have a fall back plan if things get too much for me that will allow everyone to access the models with a little more work. 

The great thing about Google Cloud Platform is that it can allow us to distribute the cost of the network traffic so that every person will pay a little bit instead of one person paying everything.  So downloading a model costs like 5  cents so the charge isn't a big deal. I can do this by making the cloud buckets a 'requester-pays' bucket. This will require people to sign up for google cloud platform  but that's all free. I don't want to do this just yet because I want the models to be as easily accessible as possible. At least for now.

**Comment by VBPXKSMI** [comment]
@Rioghasarig I know, free options are very limited in what they can do.

But I had an idea just now, say we created a torrent with all the models, would it be possible to make the Colab try to make a call to download the torrent first and see if it has enough seeders available (amount undeterminated, what is important is how fast it can be downloaded at that moment), and if the download time is under (lets say) 10 minutes, download it. Otherwise, if the download time is too long, only then call to download from Google Cloud.

This way it could save costs whenever possible, thus making it "free" for the users for a longer time.

How feasible do you think this is @machiavel23

**Comment by Rioghasarig** [comment]
Yeah, I think employing free methods and using GCS as a backup plan is definitely a good idea. 

Actually I think it'd be best if the colab didn't directly download from GCS at all. The torrents are working great for now and we should just stick with that. The GCS shouldn't be something you access every time you want to play. It's more of a backup repository if you're looking to download a model that isn't available through torrents.

**Comment by machiavel23** [comment]
> Also I don't think the colab should be set up to redownload the model every time. That'd be inefficient. There should be some permanent storage option or maybe an option to upload it from your PC if you've already downloaded it.

The colab downloads the models only on the setup phase, directly in people's gdrive. If they don't delete their gdrive before replaying, they don't redownload the model. The option to upload the model from your own computer can (and probably will) be done, but i doubt people will use it much, it'll take a ton of time for those who have poor internet and the disconnection problem might fuck it up too. It's still a reliable solution if people upload the model directly in their gdrive instead of using colab to do it

**Comment by Rioghasarig** [comment]
Oh, I see.  Yeah that sounds just great.

**Comment by Rioghasarig** [comment]
Also, @machiavel23 could you try downloading the models again? I think I fixed the issues. Let me know if it works for you. Also, speaking of long download times, you should look into "gsutil". Downloading the model with that will allow you to restart  an interrupted download. Plus there is a method I think we can use to get faster download times called "parallel composite downloads". 

EDIT: Just a heads up. I moved things around a little. The model is now stored at:
https://storage.googleapis.com/clover-edition/tensorflow/model_v5.zip
(again, Europeans should replace 'clover-edition' with 'clover-europe')

I set up a website so people could easily browse models.  (Though it may not be visible just yet). I'm probably going to be changing things  a lot for the next few days so it's probably best to just look at the website for links. 
http://www.clover-browser.info/

**Comment by machiavel23** [comment]
@VBPXKSMI  I'm not sure aria can evaluate the time needed to download a torrent before the download actually starts, it seems difficult to put in place but i get our idea.

@Rioghasarig had no luck with gsutil, probably because it's way more a hassle to [code it](https://colab.research.google.com/notebooks/snippets/accessing_files.ipynb#scrollTo=Q2CWQGIghDux) than a simple https wget,which is largely enough considering it takes only 2 minutes to grab the files, so i'll probably end up using that if i need to. Also right now, your link is 404 (yes, also with clover-europe), and the links in your website don't work with wget, but does work when i download with firefox...

There's still quite some time before I rework the model download on colab, and I'll have to check a fuckton of doc on aria to give a definitive answer, but for now, I think the end code will be grabbing from mega, then grabbing from the bucket if the user reached mega limit. people can't store more than 1 tensorflow and 1 pytorch at the same time in their gdrive anyway, so it should be largely enough for the colab setup phase

**Comment by Rioghasarig** [comment]
@machiavel23 Sorry I keep changing things up and breaking links. I moved some folders around. I'll let you know when I've settled on a scheme that should be permanent.  For now, if you go to www.clover-browser.info you can see how the directories are set up. For example, right now there is a folder called 'clover-edition' with a two files in it, model_v5_pytorch.zip and model_v5_pytorch_16-bit.zip. These files can be accessed form the bucket with the urls

https://storage.googleapis.com/clover-edition/clover-edition/model_v5_pytorch.zip
https://storage.googleapis.com/clover-edition/clover-edition/model_v5_pytorch_16-bit.zip

The first 'clover-edition' in the url is the name of the bucket. I guess this is kind of confusing. I'll probably change that. But you can get the idea. You can work out the links from the folder directory. And, I'll try and see how I can fix the set up so it's easier for you to download from it. 

And yeah, I'm kinda seeing this bucket as looking a bit in the future myself. If mega works by all means use that. I plan to train a few models though and will upload them to these buckets for people to access.

**Comment by machiavel23** [comment]
@Rioghasarig since you do it mainly for easy download for offline users, I'll get in it more deeply when I'll have finished the big colab update. I'll keep an eye on what you do though, and I'll see if I'll use it or not in colab (as a fallback solution) once you'll have finish set it up entirely. Still, if you'd like a colab test cell, or a test at a specific point, you can mention me.

**Comment by Rioghasarig** [comment]
@machiavel23. So I think I figured out how to get wget to work with the links. The problem is the link includes ampersands which the linux shell interprets incorrectly. You have to put quotes around the links. You should also use the "-O" option to save it to the right name (otherwise wget will add a bunch of extra stuff to the name of the file). 

Example:
```
wget "https://www.googleapis.com/download/storage/v1/b/clover-edition/o/test_file.txt?generation=1579989456344894&alt=media" -O test_file.txt
```
